### DISCUSSION TITLE
How can I use external OAuth2 Provider such as Gitlab or Github to replace the KeyCloak?

### DISCUSSION THREAD

**Discussion by inlann** [eclipse-theia/theia-cloud#241] [Q&A] [answered]
Hello:

I noticed that the Theia-cloud use the KeyCloak to authorize and manage users. We are new to the KeyCloak framework and the authorization flow in the KeyCloak may be diffcult to us. So, is there any way to use an external OAuth2 Provider such as Gitlab or Github to replace the KeyCloak component?

Really appreciate!

**Comment by jfaltermeier** [comment, marked as answer]
At the moment there is no way to fully replace it. 
Keycloak supports the concept of Identity providers however: https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker
With this it is possible to delegate to other social auth providers like Github, Gitlab, Google, and so on. 
https://www.keycloak.org/docs/latest/server_admin/index.html#_github
https://www.keycloak.org/docs/latest/server_admin/index.html#gitlab

**Reply by inlann** [reply]
Really appreciate!! I'll try it.

**Reply by wyhdil** [reply]
@jfaltermeier  Hello, I also tried using gitlab as Identity provider：
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/879f1128-4238-4a22-85de-b0ac655751dd)

And also configured it in gitlab：
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/1cabef8d-1031-4fda-899d-2fe754623fb9)

But when I logged in using gitlab authorization in theiacloud, an error occurred：
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/2aeb0e8f-d9b4-47a2-a984-4264ea29911a)

I don't know what causes this error. 
Is there an error in my configuration, such as an error in the callback URL in gitlab?

 Or does oauth2-proxy need some configuration to support authorized login of gitlab?
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/4d84a613-d839-4e2b-becb-e3bc29b5688c)

Can you help me how to solve this problem？Thanks!

**Reply by wyhdil** [reply]
@jfaltermeier There are some new discoveries,i added the emial option in the scope tab of gitlab.
![2729b2d7-1571-4023-af2e-7141ef43e026](https://github.com/eclipsesource/theia-cloud/assets/69418800/29f41f44-c91b-48d7-b5aa-859c3aed17b2)

Other than that, no other configuration changes were made.

But when I looked at the User page of keycloak, I found that there were some abnormalities in the account logged in using gitlab.
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/575398c8-d951-4892-a25e-6de46f72373c)

Because I knew that the Email verified switch should be turned on when configuring the correct user information, so I manually turned on the Email verified of the wyhdil account.
![ewBmBuUnFk](https://github.com/eclipsesource/theia-cloud/assets/69418800/c2eddc51-f92a-46ac-9fd0-c7d450218ff8)

Then I tried to log in using gitlab authorization and used the wyhdil account, and finally successfully opened the correct session.
![Y6FMuFMPUg](https://github.com/eclipsesource/theia-cloud/assets/69418800/9be35000-a96f-42c4-a226-896bcdcace3e)

So the above problem should be because the user’s Email is verified.
But I want to know if this is due to missing configuration? After all, it’s impossible to manually turn on the Email verified switch every time.

**Reply by jfaltermeier** [reply]
I believe that on the Gitab Identity Provider page under Advanced Settings, there is a "Trust Email" switch. 
With this all mails should be set to verified. (I am not sure what happens with existing users though)

**Reply by wyhdil** [reply]
Ok,this really works!!!! Thank you soooooo much!
